---
layout:     post
title:      MySQL报表从库突然宕机过程回顾和原因分析.md
subtitle:  	
date:       2020-03-21
author:     RC
header-img: 
catalog: true
tags:
    - MySQL
    - crash and recovery
    - 故障排查
---

## 1.过程回顾

3月18日（周三）下午16点59分左右，DBA收到webhook转发的报警邮件，内容为MySQL数据库无法连接：

![1](https://i.postimg.cc/vHdsQ0Y2/11.png)

由于考拉的这台报表数据库所在的服务器是一台物理机上的虚拟机，因此一开始认为是资源争抢导致监控延迟，但是登录机器后发现MySQL进程已经宕掉。正在使用的开发也反映无法连接到该报表数据库。

尝试正常重启MySQL进程服务，数次后依然失败。

观察错误日志，发现可能与MySQL key_buffer_size等参数过大有关：

![1](https://i.postimg.cc/fL0Gs96h/22.png)

此时key_buffer_size设置的是8G
尝试将该参数和其他参数均减小到一定程度，正常启动MySQL进程服务依然失败，错误日志抛出的信息不变。

不得已将innodb_force_recovery设为最高级别写入配置文件，启动成功，**但是主从复制无法进行，任何表均无法访问，显示cannot find 所有表的tablespace**

商量决定通过当日的全量备份还原报表库，因为该从库上还有一个reportteam的开发单独使用库，因此通过mysqldump导出该库，在全备还原完成和主从搭建后再还原回去。

通过xtrabackup还原data相关文件完成后，重启MySQL进程，依然报相同错误，强制启动后，错误日志输出和还原前一致。

这里可以判断已经不是ibdata1共享表空间文件或者其他ibd文件以及日志文件损坏造成的无法正常启动了，问题比较严重。

因为手头还有数据的备份，因此不得已选择重新安装MySQL，再次还原，建立主从，还原单独库。

最终在晚上19点15分左右全部还原成功，到今天目前观察正常。


## 2.故障问题分析

我们再来看下参数设置过大的报错内容的另外一部分：

![1](https://i.postimg.cc/ZKJM72Km/33.png)

可以看到MySQL官方对这个crash宕机后无法启动的问题也是心知肚明，只言片语中提到的表空间损坏也与强行启动后错误日志反映的一致。

求助于谷歌，发现MySQL官方的bug讨论串里面，虽然现象差不多，但是反应的都是比较老的版本（5.6以下），
并且有人提到5.7版本已经修复：
https://bugs.mysql.com/bug.php?id=61516
而国内的一些案例分享中，也只是到还原备份那一步，或者删除共享表空间文件重新还原后就可以正常work了，并没有这次情况这么极端。

**在错误日志中我们还可以看到MySQL给出了程序报错的具体行数**，

这里参考下5.6的MySQL源码相关部分：

![1](https://i.postimg.cc/520k1zCk/44.png)

可以看到**报错的168行**是一个类似于Python语法中的**断言（assert）**，

根据该断言的定义来看，

![1](https://i.postimg.cc/pd5G5tkf/55.png)

![1](https://i.postimg.cc/HxGSQPPf/66.png)

**只要断言中的判定失败，那么就会断言错误，并且中断程序执行**，

这个错误就是错误日志中的:

![1](https://i.postimg.cc/VvmDh0wg/77.png)

再看一下断言中的判断内容：

![1](https://i.postimg.cc/9Qr1sj66/88.png)

![1](https://i.postimg.cc/NjSbczY5/99.png)

![1](https://i.postimg.cc/xCyPLjmz/1010.png)

![1](https://i.postimg.cc/sfmcXXsF/1111.png)

我们知道，MySQL因为使用了MVCC机制，在进行记录的删除和更新操作时，原来的记录的delete mark会被置为1，而不是直接删除数据，为了回收过期的undo页和清除已经被mark的数据，使用了purge线程来操作。而数据库的数据页很多，要清除被删除的数据，不可能遍历所有的数据页。由于所有的变更都有undo log， 因此，从undo作为切入点，在清理过期的undo的同时，也将数据页中的被删除的记录一并清除。

而默认的Innodb_fast_shutdown=1，意味着每次关机时不会进行purge的操作，**那么这个操作会在启动时进行，**

而在启动过程中的purge操作时，

![1](https://i.postimg.cc/d3VGd2vb/1212.png)

![1](https://i.postimg.cc/6qbnh8py/1313.png)

![1](https://i.postimg.cc/7YbgZnK7/1414.png)

正常情况下，需要确定purge_sys->iter是不能大于purge_sys->limit的，原因是由于purge->limit是用来truncate对应的undo log并且更新history list，而iter用于找到对应del的data record进行purge，我们一定要保证purge del的data后才能purge对应的undo log。

在这次故障抛出的错误中可以看到，purge_sys->iter的trx_no，即**指定回滚段上最早提交的事务号，已经比最老的读事务的事务提交号还要大，意味着purge线程在del部分标记记录后，并没有清除对应的过期undo log，因此获得的事务号超前，导致无法完成purge线程的内容，最终抛错无法启动**。

>(这部分个人理解因为水平有限可能不准确)

## 3.小结

可以看到，从各种故障迅速恢复的手段还是依赖于备份和高可用，横向扩展在一定程度上可以防止意外宕机造成的业务损失，但是备份依然高于一切，如何做好有效的备份，并在第一时间恢复，是必须要重视的。
